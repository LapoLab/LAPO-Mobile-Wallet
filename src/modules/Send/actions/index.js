// @flow
import { createActions } from 'redux-actions';

export const {
  send, sendSuccess, sendError, setSendAddress, setSendAmount, setSendNote,
} = createActions({
  SEND: (address = '', amount = 0, note = '') => ({
    address,
    amount,
    note,
  }),
  SEND_SUCCESS: (data = '') => data,
  SEND_ERROR: (data = '') => data,
  SET_SEND_ADDRESS: (data = '') => data,
  SET_SEND_AMOUNT: (data = '') => data,
  SET_SEND_NOTE: (data = '') => data,
});
