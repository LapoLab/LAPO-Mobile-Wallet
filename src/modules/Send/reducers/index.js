import { handleActions } from 'redux-actions';
import {
  setSendAddress,
  setSendAmount,
  setSendNote,
} from '../actions';

export const initialSendState = {
  address: '',
  amount: '',
  note: '',
};

// reducer
const reducer = handleActions(
  {
    [setSendAddress](state, { payload }) {
      return {
        ...state,
        address: payload,
      };
    },
    [setSendAmount](state, { payload }) {
      return {
        ...state,
        amount: payload,
      };
    },
    [setSendNote](state, { payload }) {
      return {
        ...state,
        note: payload,
      };
    },
  },
  initialSendState,
);

export default reducer;
