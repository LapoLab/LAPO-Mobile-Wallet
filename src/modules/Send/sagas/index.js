import send from './sendSaga';

export * from '../actions';

export default {
  send,
};
