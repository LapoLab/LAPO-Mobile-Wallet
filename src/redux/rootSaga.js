import { all } from 'redux-saga/effects';
import commonsSagas from '../commons/sagas';
import recoveryWordsSagas from '../modules/RecoveryWords/sagas';
import sendSagas from '../modules/Send/sagas';
import paperWordsConfirm from '../modules/PaperWordsConfirm/sagas';
import dashboard from '../modules/Dashboard/sagas';
import settings from '../modules/Settings/sagas';
import paperWords from '../modules/PaperWords/sagas';
import transactionsDetails from '../modules/TransactionDetails/sagas';
import chooseContact from '../modules/ChooseContact/sagas';
import splashScreen from '../modules/SplashScreen/sagas';
import switchContacts from '../modules/SwitchContacts/sagas';
import receive from '../modules/Receive/sagas';

// eslint-disable-next-line
const getListeners = (...args) => args.reduce((acc, nextArg) => [...acc, ...Object.values(nextArg).map(func => func())], []);

export default function* rootSaga() {
  yield all(
    getListeners(
      commonsSagas,
      recoveryWordsSagas,
      sendSagas,
      paperWordsConfirm,
      dashboard,
      settings,
      paperWords,
      transactionsDetails,
      chooseContact,
      splashScreen,
      switchContacts,
      receive,
    ),
  );
}
