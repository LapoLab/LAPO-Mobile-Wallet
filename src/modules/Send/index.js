import Send from './component';

export default Send;
