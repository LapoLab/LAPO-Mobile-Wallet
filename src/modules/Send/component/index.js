/**
 * @flow
 */

import React, { Component } from 'react';
import {
  Text,
  TextInput,
  View,
  TouchableHighlight,
  Image,
  TouchableOpacity,
  Clipboard,
  Platform,
  KeyboardAvoidingView,
  Keyboard,
} from 'react-native';
import { connect } from 'react-redux';
import { NavigationActions } from 'react-navigation';
import { bindActionCreators } from 'redux';
import * as actions from '../actions';
import * as actionsCommon from '../../../commons/actions';
import SwipeButton from '../../../commons/components/SwipeButton';
import SpinnerPage from '../../../commons/components/SpinnerPage';
import styles from './styles';
import {
  close, qr, Users2Icon, copyIcon,
} from '../../../assets';

type Props = {
  navigation: { dispatch: Function, navigate: Function },
  onClose: Function,
  returnDataQrCode: Function,
  send: Function,
  address: string,
  setAddress: Function,
  setAmount: Function,
  setNote: Function,
  goToOtherWindow: Function,
  address: string,
  amount: string,
  note: string,
};
type State = {
  isShowKeyboard: boolean,
};
export class SendModal extends Component<Props, State> {
  keyboardDidShowListener: Function;

  keyboardDidHideListener: Function;

  state = {
    isShowKeyboard: false,
  };

  componentDidMount() {
    this.keyboardDidShowListener = Keyboard.addListener('keyboardDidShow', this.keyboardDidShow);
    this.keyboardDidHideListener = Keyboard.addListener('keyboardDidHide', this.keyboardDidHide);
  }

  componentWillUnmount() {
    this.keyboardDidShowListener.remove();
    this.keyboardDidHideListener.remove();
  }

  keyboardDidShow = () => {
    this.setState({ isShowKeyboard: true });
  };

  keyboardDidHide = () => {
    this.setState({ isShowKeyboard: false });
  };

  navigateToScreen = (route: string) => () => {
    const {
      navigation: { dispatch },
    } = this.props;
    const navigateAction = NavigationActions.navigate({
      routeName: route,
    });
    dispatch(navigateAction);
  };

  handleChooseContact = () => {
    const {
      navigation: { navigate },
      onClose,
    } = this.props;
    onClose();
    navigate('ChooseContact');
  };

  handleClose = () => {
    const { onClose } = this.props;
    onClose();
  };

  handleSend = () => {
    const {
      send, amount, address, note,
    } = this.props;

    send(address, amount, note);
  };

  handleChangeAddress = (e: string) => {
    const { setAddress } = this.props;
    setAddress(e);
  };

  handleChangeAmount = (e: string) => {
    const { setAmount, amount } = this.props;
    if (e.length === 1 && e === '0' && amount.length === 0) {
      setAmount('0.');
    } else {
      setAmount(
        e
          .replace(/,/g, '.')
          .replace(/,,/g, '.')
          .replace(/\.\./g, '.'),
      );
    }
  };

  handleChangeNote = (e: string) => {
    const { setNote } = this.props;
    setNote(e);
  };

  handlePaste = async () => {
    const { setAddress } = this.props;
    const content = await Clipboard.getString();
    setAddress(content || '');
  };

  handleScanQrCode = () => {
    const {
      navigation: { navigate },
      onClose,
      returnDataQrCode,
      setAddress,
      goToOtherWindow,
    } = this.props;
    goToOtherWindow(true);
    setAddress('');
    onClose();
    navigate('QrCodeScanSend', { returnData: returnDataQrCode });
  };

  handleCloseKeyboard = () => {
    const { isShowKeyboard } = this.state;
    if (isShowKeyboard) {
      Keyboard.dismiss();
    }
  };

  render() {
    const {
      address, amount, note,
    } = this.props;
    return (
      <KeyboardAvoidingView
        style={styles.keyboard}
        behavior={Platform.OS === 'ios' ? 'padding' : ''}
        enabled
      >
        <TouchableHighlight
          style={styles.GlobalContainer}
          onPress={this.handleClose}
          underlayColor="transparent"
        >
          <TouchableHighlight
            style={styles.container}
            onPress={this.handleCloseKeyboard}
            underlayColor="#EFF4FC"
          >
            <View style={{ flex: 1, width: '100%' }}>
              <View style={styles.header}>
                <TouchableOpacity onPress={this.handleClose} style={styles.buttonClose}>
                  <Image style={styles.headerIcon} source={close} />
                </TouchableOpacity>
                <Text style={styles.headerTitle}>Send</Text>
                <TouchableOpacity onPress={this.handleScanQrCode}>
                  <Image style={styles.headerIconRight} source={qr} />
                </TouchableOpacity>
              </View>

              <View style={styles.row}>
                <Text style={styles.rowTitle}>To</Text>
                <TextInput
                  style={styles.valueAddress}
                  value={address}
                  autoCapitalize="none"
                  onChangeText={this.handleChangeAddress}
                />
                <TouchableOpacity style={styles.buttonRowUser} onPress={this.handlePaste}>
                  <Image style={styles.copyIcon} source={copyIcon} />
                </TouchableOpacity>
                <TouchableOpacity style={styles.buttonRowUser} onPress={this.handleChooseContact}>
                  <Image style={styles.users2Icon} source={Users2Icon} />
                </TouchableOpacity>
              </View>

              <View style={styles.row}>
                <Text style={styles.rowTitle}>Amount</Text>
                <TextInput
                  style={styles.valueAddress}
                  keyboardType="numeric"
                  value={amount}
                  onChangeText={this.handleChangeAmount}
                />
                <TouchableOpacity style={styles.buttonRow}>
                  <Text style={styles.textButton}>LAX</Text>
                </TouchableOpacity>
              </View>

              <View style={styles.row}>
                <Text style={styles.rowTitle}>Note</Text>
                <TextInput
                  style={styles.valueAddress}
                  value={note}
                  onChangeText={this.handleChangeNote}
                />
              </View>

              <View style={styles.buttonContainer}>
                <SwipeButton onUnlock={this.handleSend} />
              </View>
            </View>
          </TouchableHighlight>
        </TouchableHighlight>

        <SpinnerPage />
      </KeyboardAvoidingView>
    );
  }
}

const mapStateToProps = state => ({
  address: state.send.address,
  amount: state.send.amount,
  note: state.send.note,
});

const mapDispatchToProps = dispatch => bindActionCreators(
  {
    send: actions.send,
    setAddress: actions.setSendAddress,
    setAmount: actions.setSendAmount,
    setNote: actions.setSendNote,
    goToOtherWindow: actionsCommon.goToOtherWindow,
  },
  dispatch,
);

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(SendModal);
