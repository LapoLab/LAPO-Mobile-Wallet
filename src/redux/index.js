import { createStore, applyMiddleware, compose } from 'redux';
import createSagaMiddleware from 'redux-saga';
import storage from 'redux-persist/lib/storage';
import { persistStore, persistReducer } from 'redux-persist';
import autoMergeLevel2 from 'redux-persist/lib/stateReconciler/autoMergeLevel2';
import rootReducer from './rootReducer';
import rootSagas from './rootSaga';

const persistConfig = {
  key: 'root',
  storage,
  stateReconciler: autoMergeLevel2,
  blacklist: [
    'modals',
    'dashboard',
    'paperWords',
    'chooseContact',
    'send',
    'transactionsDetails',
    'alert',
    'pinsRecovery',
  ],
};

const persistedReducer = persistReducer(persistConfig, rootReducer);

const sagaMiddleware = createSagaMiddleware();
// eslint-disable-next-line no-undef
const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

export const store = createStore(
  persistedReducer,
  composeEnhancers(applyMiddleware(sagaMiddleware)),
);
sagaMiddleware.run(rootSagas);

export const persistor = persistStore(store);
