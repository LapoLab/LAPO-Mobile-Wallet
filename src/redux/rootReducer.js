import { combineReducers } from 'redux';

import wallet from '../commons/reducers/wallet';
import pins from '../commons/reducers/pins';
import pinsRecovery from '../commons/reducers/pinsRecovery';
import modals from '../commons/reducers/modals';
import settings from '../commons/reducers/settings';
import notify from '../modules/Notify/reducers';
import paperConfirm from '../modules/PaperWordsConfirm/reducer';
import dashboard from '../modules/Dashboard/reducers';
import paperWords from '../modules/PaperWords/reducers';
import chooseContact from '../modules/ChooseContact/reducers';
import send from '../modules/Send/reducers';
import transactionsDetails from '../modules/TransactionDetails/reducer';
import alert from '../modules/Alert/reducers';

export default combineReducers({
  wallet,
  pins,
  pinsRecovery,
  notify,
  modals,
  settings,
  paperConfirm,
  dashboard,
  paperWords,
  chooseContact,
  send,
  transactionsDetails,
  alert,
});
