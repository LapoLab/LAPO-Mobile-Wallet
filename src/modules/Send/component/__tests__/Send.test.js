// @flow
import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { store } from '../../../../redux';

import Send from '../index';


describe('Send modal', () => {
  const tree = renderer.create(
    <Provider store={store}>
      <Send />
    </Provider>,
  );

  const render = tree.toJSON();

  test('renders snapshots', () => {
    expect(render).toMatchSnapshot();
  });
});
