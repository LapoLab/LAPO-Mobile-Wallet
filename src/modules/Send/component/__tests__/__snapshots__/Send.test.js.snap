// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Send modal renders snapshots 1`] = `
<View
  onLayout={[Function]}
  style={
    Object {
      "flex": 1,
      "justifyContent": "space-between",
    }
  }
>
  <View
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "transparent",
        "flex": 1,
        "flexDirection": "column",
        "justifyContent": "flex-end",
      }
    }
  >
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "flex-start",
          "backgroundColor": "#EFF4FC",
          "borderTopLeftRadius": 12,
          "borderTopRightRadius": 12,
          "height": 350,
          "width": "100%",
        }
      }
    >
      <View
        style={
          Object {
            "flex": 1,
            "width": "100%",
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "borderBottomColor": "#D8D8D8",
              "borderBottomWidth": 1,
              "flexDirection": "row",
              "height": 60,
              "justifyContent": "space-between",
              "paddingRight": 15,
              "paddingVertical": 15,
            }
          }
        >
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
                "padding": 15,
              }
            }
          >
            <Image
              source={
                Object {
                  "testUri": "../../../src/assets/close.png",
                }
              }
              style={
                Object {
                  "height": 15,
                  "resizeMode": "contain",
                  "width": 15,
                }
              }
            />
          </View>
          <Text
            style={
              Object {
                "color": "#131313",
                "fontFamily": "Lato-Bold",
                "fontSize": 18,
                "marginRight": 5,
              }
            }
          >
            Send
          </Text>
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <Image
              source={
                Object {
                  "testUri": "../../../src/assets/qr.png",
                }
              }
              style={
                Object {
                  "height": 20,
                  "marginTop": 2,
                  "resizeMode": "contain",
                  "width": 20,
                }
              }
            />
          </View>
        </View>
        <View
          style={
            Object {
              "alignItems": "center",
              "borderBottomColor": "#D8D8D8",
              "borderBottomWidth": 1,
              "flexDirection": "row",
              "height": 60,
              "justifyContent": "space-between",
              "paddingHorizontal": 15,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#565656",
                "fontFamily": "Lato-Bold",
                "fontSize": 16,
                "lineHeight": 24,
                "width": 65,
              }
            }
          >
            To
          </Text>
          <TextInput
            allowFontScaling={true}
            autoCapitalize="none"
            onChangeText={[Function]}
            style={
              Object {
                "color": "black",
                "flex": 1,
                "fontFamily": "Lato-Bold",
                "fontSize": 16,
                "marginBottom": 6,
                "marginLeft": 10,
                "marginTop": 8,
                "textAlign": "left",
              }
            }
            underlineColorAndroid="transparent"
            value=""
          />
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "backgroundColor": "#BBCDEB",
                "borderRadius": 5,
                "justifyContent": "center",
                "marginLeft": 10,
                "opacity": 1,
                "paddingHorizontal": 10,
                "paddingVertical": 10,
              }
            }
          >
            <Image
              source={
                Object {
                  "testUri": "../../../src/assets/copyIcon.png",
                }
              }
              style={
                Object {
                  "height": 15,
                  "resizeMode": "contain",
                  "width": 15,
                }
              }
            />
          </View>
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "backgroundColor": "#BBCDEB",
                "borderRadius": 5,
                "justifyContent": "center",
                "marginLeft": 10,
                "opacity": 1,
                "paddingHorizontal": 10,
                "paddingVertical": 10,
              }
            }
          >
            <Image
              source={
                Object {
                  "testUri": "../../../src/assets/users-2-icon.png",
                }
              }
              style={
                Object {
                  "height": 15,
                  "resizeMode": "contain",
                  "width": 15,
                }
              }
            />
          </View>
        </View>
        <View
          style={
            Object {
              "alignItems": "center",
              "borderBottomColor": "#D8D8D8",
              "borderBottomWidth": 1,
              "flexDirection": "row",
              "height": 60,
              "justifyContent": "space-between",
              "paddingHorizontal": 15,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#565656",
                "fontFamily": "Lato-Bold",
                "fontSize": 16,
                "lineHeight": 24,
                "width": 65,
              }
            }
          >
            Amount
          </Text>
          <TextInput
            allowFontScaling={true}
            keyboardType="numeric"
            onChangeText={[Function]}
            style={
              Object {
                "color": "black",
                "flex": 1,
                "fontFamily": "Lato-Bold",
                "fontSize": 16,
                "marginBottom": 6,
                "marginLeft": 10,
                "marginTop": 8,
                "textAlign": "left",
              }
            }
            underlineColorAndroid="transparent"
            value=""
          />
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "backgroundColor": "#BBCDEB",
                "borderRadius": 5,
                "fontFamily": "Lato-Bold",
                "justifyContent": "center",
                "minWidth": 60,
                "opacity": 1,
                "paddingHorizontal": 10,
                "paddingVertical": 10,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "black",
                  "fontFamily": "Lato-Light",
                  "fontSize": 14,
                  "textAlign": "center",
                }
              }
            >
              LAX
            </Text>
          </View>
        </View>
        <View
          style={
            Object {
              "alignItems": "center",
              "borderBottomColor": "#D8D8D8",
              "borderBottomWidth": 1,
              "flexDirection": "row",
              "height": 60,
              "justifyContent": "space-between",
              "paddingHorizontal": 15,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#565656",
                "fontFamily": "Lato-Bold",
                "fontSize": 16,
                "lineHeight": 24,
                "width": 65,
              }
            }
          >
            Note
          </Text>
          <TextInput
            allowFontScaling={true}
            onChangeText={[Function]}
            style={
              Object {
                "color": "black",
                "flex": 1,
                "fontFamily": "Lato-Bold",
                "fontSize": 16,
                "marginBottom": 6,
                "marginLeft": 10,
                "marginTop": 8,
                "textAlign": "left",
              }
            }
            underlineColorAndroid="transparent"
            value=""
          />
        </View>
        <View
          style={
            Object {
              "alignItems": "center",
              "flex": 1,
              "justifyContent": "center",
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#BBCDEB",
                "borderRadius": 30,
                "flexDirection": "row",
                "height": 64,
                "justifyContent": "flex-start",
                "paddingHorizontal": 10,
                "width": "80%",
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "height": 56,
                  "position": "absolute",
                  "top": 22,
                  "width": 600,
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "white",
                    "fontFamily": "Lato-Regular",
                    "fontSize": 18,
                  }
                }
              >
                SLIDE TO SEND
              </Text>
            </View>
            <View
              onMoveShouldSetResponder={[Function]}
              onMoveShouldSetResponderCapture={[Function]}
              onResponderEnd={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderReject={[Function]}
              onResponderRelease={[Function]}
              onResponderStart={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              onStartShouldSetResponderCapture={[Function]}
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#4183BD",
                  "borderRadius": 26,
                  "height": 52,
                  "justifyContent": "center",
                  "marginLeft": "-1%",
                  "width": 52,
                }
              }
            >
              <Image
                source={
                  Object {
                    "testUri": "../../../src/assets/sendButton.png",
                  }
                }
                style={
                  Object {
                    "resizeMode": "contain",
                    "width": 24,
                  }
                }
              />
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
